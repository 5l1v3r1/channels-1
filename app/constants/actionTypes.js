// @flow
const actionTypes = {
  FETCH_ACCESS_TOKEN_REQUEST: 'FETCH_ACCESS_TOKEN_REQUEST',
  FETCH_ACCESS_TOKEN_SUCCESS: 'FETCH_ACCESS_TOKEN_SUCCESS',
  FETCH_ACCESS_TOKEN_FAILURE: 'FETCH_ACCESS_TOKEN_FAILURE',
  FETCH_CHANNELS_REQUEST: 'FETCH_CHANNELS_REQUEST',
  FETCH_CHANNELS_SUCCESS: 'FETCH_CHANNELS_SUCCESS',
  FETCH_CHANNELS_FAILURE: 'FETCH_CHANNELS_FAILURE',
  UPDATE_ACTIVE_CHANNEL: 'UPDATE_ACTIVE_CHANNEL',
  FETCH_VIDEOS_REQUEST: 'FETCH_VIDEOS_REQUEST',
  FETCH_VIDEOS_SUCCESS: 'FETCH_VIDEOS_SUCCESS',
  FETCH_VIDEOS_FAILURE: 'FETCH_VIDEOS_FAILURE'
};

export default actionTypes;
