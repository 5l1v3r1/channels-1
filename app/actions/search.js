// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, ThunkAction, Dispatch, GetState } from '../constants/types';
import type { VideoItem } from '../dataTypes/videoType';
import { getRequest } from '../services/fetch';
import { getVideosUri, getVideoIdsUri } from '../services/uriGenerator';

export function fetchSearch(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(fetchSearchRequest());
    const params = {
      channelId: getState().channels.activeId,
      q: getState().search.query,
      pageToken: getState().search.pageToken,
      order: 'relevance'
    };
    return fetchVideoIds(params)
      .then(
        res => {
          const { videoIds, nextPageToken } = res;
          const uri = getVideosUri(videoIds);
          return getRequest(uri)
            .then(
              json => dispatch(fetchSearchSuccess(json.items, nextPageToken)),
              () => dispatch(fetchSearchFailure())
            );
        },
        () => dispatch(fetchSearchFailure())
      );
  };
}

export function fetchVideoIds(params: any): Promise<any> {
  const uri = getVideoIdsUri(params);
  return getRequest(uri)
    .then(
      json => Promise.resolve({
        videoIds: json.items.map(item => item.id.videoId).join(),
        nextPageToken: json.nextPageToken || params.pageToken
      }),
      () => Promise.reject()
    );
}

export function fetchSearchRequest(): Action {
  return {
    type: actionTypes.FETCH_SEARCH_REQUEST
  };
}

export function fetchSearchSuccess(items: VideoItem[], nextPageToken: string): Action {
  return {
    type: actionTypes.FETCH_SEARCH_SUCCESS,
    payload: {
      items,
      nextPageToken
    }
  };
}

export function fetchSearchFailure(): Action {
  return {
    type: actionTypes.FETCH_SEARCH_FAILURE
  };
}

export function updateSearch(query: string): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(clearSearchResults());
    dispatch(updateSearchQuery(query));
  };
}

export function clearSearch(): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(updateSearchQuery(''));
    dispatch(clearSearchResults());
  };
}

export function updateSearchQuery(query: string): Action {
  return {
    type: actionTypes.UPDATE_SEARCH_QUERY,
    payload: query
  };
}

export function clearSearchResults(): Action {
  return {
    type: actionTypes.CLEAR_SEARCH_RESULTS
  };
}
