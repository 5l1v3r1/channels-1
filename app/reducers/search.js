// @flow
import { combineReducers } from 'redux';
import actionTypes from '../constants/actionTypes';
import type { Action } from '../constants/typeAliases';
import VideoType from '../dataTypes/videoType';
import { createIsFetching, createError } from './common';

function query(state: string = '', { type, payload }: Action): string {
  switch (type) {
    case actionTypes.UPDATE_SEARCH_QUERY:
      return payload;
    default:
      return state;
  }
}

function results(state: VideoType[] = [], { type, payload }: Action): VideoType[] {
  switch (type) {
    case actionTypes.FETCH_SEARCH_SUCCESS:
      return [
        ...state,
        ...payload.items.map(item => VideoType.from(item))
      ];
    case actionTypes.CLEAR_SEARCH_RESULTS:
      return [];
    default:
      return state;
  }
}

function pageToken(state: string = '', { type, payload }: Action): string {
  switch (type) {
    case actionTypes.FETCH_SEARCH_SUCCESS:
      return payload.nextPageToken;
    case actionTypes.CLEAR_SEARCH_RESULTS:
      return '';
    default:
      return state;
  }
}

const search = combineReducers({
  query,
  results,
  pageToken,
  isFetching: createIsFetching('SEARCH'),
  error: createError('SEARCH')
});

export default search;
