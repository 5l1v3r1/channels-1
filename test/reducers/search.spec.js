import actionTypes from '../../app/constants/actionTypes';
import search from '../../app/reducers/search';
import VideoType from '../../app/dataTypes/videoType';

describe('search reducer', () => {
  it('should return initial state', () => {
    expect(search(undefined, {})).toEqual({
      query: '',
      results: [],
      pageToken: '',
      isFetching: false,
      error: false
    });
  });

  it('should handle FETCH_SEARCH_SUCCESS', () => {
    VideoType.from = jest.fn(item => item);

    expect(search(
      {
        query: 'react',
        results: [],
        pageToken: '',
        isFetching: false,
        error: false
      },
      {
        type: actionTypes.FETCH_SEARCH_SUCCESS,
        payload: {
          items: [
            { id: '1' },
            { id: '2' }
          ],
          nextPageToken: 'token'
        }
      }
    )).toEqual({
      query: 'react',
      results: [
        { id: '1' },
        { id: '2' }
      ],
      pageToken: 'token',
      isFetching: false,
      error: false
    });

    expect(search(
      {
        query: 'react',
        results: [{ id: '1' }],
        pageToken: 'token',
        isFetching: false,
        error: false
      },
      {
        type: actionTypes.FETCH_SEARCH_SUCCESS,
        payload: {
          items: [{ id: '2' }],
          nextPageToken: 'next token'
        }
      }
    )).toEqual({
      query: 'react',
      results: [
        { id: '1' },
        { id: '2' }
      ],
      pageToken: 'next token',
      isFetching: false,
      error: false
    });
  });

  it('should handle UPDATE_SEARCH_QUERY', () => {
    expect(search(
      undefined,
      {
        type: actionTypes.UPDATE_SEARCH_QUERY,
        payload: 'x'
      }
    )).toEqual({
      query: 'x',
      results: [],
      pageToken: '',
      isFetching: false,
      error: false
    });

    expect(search(
      {
        query: 'x',
        results: [],
        pageToken: '',
        isFetching: false,
        error: false
      },
      {
        type: actionTypes.UPDATE_SEARCH_QUERY,
        payload: 'xo'
      }
    )).toEqual({
      query: 'xo',
      results: [],
      pageToken: '',
      isFetching: false,
      error: false
    });
  });

  it('should handle CLEAR_SEARCH_RESULTS', () => {
    const state = {
      query: 'react',
      results: [
        { id: '1' },
        { id: '2' }
      ],
      pageToken: 'token',
      isFetching: false,
      error: false
    };
    const action = { type: actionTypes.CLEAR_SEARCH_RESULTS };

    expect(search(state, action)).toEqual({
      query: 'react',
      results: [],
      pageToken: '',
      isFetching: false,
      error: false
    });
  });

  it('should handle an unknown action type', () => {
    const action = { type: 'UNKNOWN' };

    expect(search(undefined, action)).toEqual({
      query: '',
      results: [],
      pageToken: '',
      isFetching: false,
      error: false
    });
  });
});
