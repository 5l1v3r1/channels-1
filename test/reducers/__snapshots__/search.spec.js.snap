// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search reducer should handle CLEAR_SEARCH_RESULTS 1`] = `
Object {
  "error": false,
  "isFetching": false,
  "pageToken": "",
  "query": "abc",
  "results": Array [],
}
`;

exports[`search reducer should handle FETCH_SEARCH_FAILURE 1`] = `
Object {
  "error": true,
  "isFetching": false,
  "pageToken": "",
  "query": "a",
  "results": Array [],
}
`;

exports[`search reducer should handle FETCH_SEARCH_REQUEST 1`] = `
Object {
  "error": false,
  "isFetching": true,
  "pageToken": "",
  "query": "",
  "results": Array [],
}
`;

exports[`search reducer should handle FETCH_SEARCH_SUCCESS 1`] = `
Object {
  "error": false,
  "isFetching": false,
  "pageToken": "CBQQAA",
  "query": "a",
  "results": Array [
    Object {
      "id": "M8l2aGMjKHI",
    },
    Object {
      "id": "GcSACxUbqtg",
    },
  ],
}
`;

exports[`search reducer should handle UPDATE_SEARCH_QUERY 1`] = `
Object {
  "error": false,
  "isFetching": false,
  "pageToken": "",
  "query": "a",
  "results": Array [],
}
`;

exports[`search reducer should handle UPDATE_SEARCH_QUERY 2`] = `
Object {
  "error": false,
  "isFetching": false,
  "pageToken": "",
  "query": "ab",
  "results": Array [],
}
`;

exports[`search reducer should handle unknown action type 1`] = `
Object {
  "error": false,
  "isFetching": false,
  "pageToken": "",
  "query": "",
  "results": Array [],
}
`;

exports[`search reducer should return initial state 1`] = `
Object {
  "error": false,
  "isFetching": false,
  "pageToken": "",
  "query": "",
  "results": Array [],
}
`;
